### PR TITLE
AMCL: Set an initial guess by service call

### PR DESCRIPTION
**Pull request by aock**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | None |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | gazebo simulation |
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points

One can send initial pose guesses via services now. I closely oriented at the already existing "/reinitialize_global_localization" service but I additionally had to create a new `SetInitialPose` service interface in `nav2_msgs`. In my tests, setting the pose and the covariances had the same behavior as using the topic "/initialpose". 

## Description of documentation updates required from your changes

I added a new service "/set_initial_pose". So I guess that could be added to the list of services in the documentation. 

---

## Future work that may be required in bullet points

none

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

**Comment by SteveMacenski** [comment]
@aock please explain the need for this feature over the topic interface that already exists

**Comment by aock** [comment]
Hi @SteveMacenski 

It's more about the concepts. I am well aware of reliable communication in ROS2 with topics. However, according to the [docs](https://docs.ros.org/en/foxy/How-To-Guides/Topics-Services-Actions.html), topics should be used for "continuous data streams". And services for "remote procedure calls that end quickly". And that is the case with this one. Just like for the "/global_localization" service. Nevertheless, I would still keep the topic interface for RViz.

-- Alex

File: nav2_amcl/include/nav2_amcl/amcl_node.hpp [hunk 1]
```diff
@@ -223,6 +234,7 @@ class AmclNode : public nav2_util::LifecycleNode
   // Nomotion update control. Used to temporarily let amcl update samples even when no motion occurs
   std::atomic<bool> force_update_{false};
 
+
```
Comment (SteveMacenski) [inline nav2_amcl/include/nav2_amcl/amcl_node.hpp:237]: Remove empty line added

File: nav2_amcl/include/nav2_amcl/amcl_node.hpp [hunk 2]
```diff
@@ -210,6 +211,16 @@ class AmclNode : public nav2_util::LifecycleNode
     const std::shared_ptr<std_srvs::srv::Empty::Request> request,
     std::shared_ptr<std_srvs::srv::Empty::Response> response);
 
+  // service server for providing an initial pose guess
+  rclcpp::Service<nav2_msgs::srv::SetInitialPose>::SharedPtr initial_guess_srv_;
+  /*
+   * @brief Service callback for a global relocalization request
+   */
+  void initialGuessCallback(
```
Comment (SteveMacenski) [inline nav2_amcl/include/nav2_amcl/amcl_node.hpp:219]: `initialPoseReceivedSrv`

File: nav2_msgs/srv/SetInitialPose.srv [hunk 3]
```diff
@@ -0,0 +1,2 @@
+geometry_msgs/PoseWithCovarianceStamped pose
+---
```
Comment (SteveMacenski) [inline nav2_msgs/srv/SetInitialPose.srv:2]: Missing new line

File: nav2_amcl/src/amcl_node.cpp [hunk 4]
```diff
@@ -494,6 +494,16 @@ AmclNode::globalLocalizationCallback(
...
 }
 
+void
+AmclNode::initialGuessCallback(
+  const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+  const std::shared_ptr<nav2_msgs::srv::SetInitialPose::Request> req,
+  std::shared_ptr<nav2_msgs::srv::SetInitialPose::Response>/*res*/)
+{
+  std::lock_guard<std::recursive_mutex> cfl(mutex_);
+  initialPoseReceived(std::make_shared<geometry_msgs::msg::PoseWithCovarianceStamped>(req->pose));
```
Comment (SteveMacenski) [inline nav2_amcl/src/amcl_node.cpp:504]: `initialPoseReceived` locks the mutex, no need to do that here

File: nav2_amcl/src/amcl_node.cpp [hunk 5]
```diff
@@ -1543,6 +1553,10 @@ AmclNode::initServices()
     "reinitialize_global_localization",
     std::bind(&AmclNode::globalLocalizationCallback, this, _1, _2, _3));
 
+  initial_guess_srv_ = create_service<nav2_msgs::srv::SetInitialPose>(
```
Comment (SteveMacenski) [inline nav2_amcl/src/amcl_node.cpp:1556]: This must be reset on cleanup like the others

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

**Comment by SteveMacenski** [comment]
also, heed the DCO requirement in the CI badges

File: nav2_amcl/include/nav2_amcl/amcl_node.hpp (hunk 1 above)
Comment (aock) [reply nav2_amcl/include/nav2_amcl/amcl_node.hpp]: Removed.

File: nav2_amcl/include/nav2_amcl/amcl_node.hpp (hunk 2 above)
Comment (aock) [reply nav2_amcl/include/nav2_amcl/amcl_node.hpp]: Renamed the callback name to initialPoseReceivedSrv.

File: nav2_msgs/srv/SetInitialPose.srv (hunk 3 above)
Comment (aock) [reply nav2_msgs/srv/SetInitialPose.srv]: Added missing new line.

File: nav2_amcl/src/amcl_node.cpp (hunk 4 above)
Comment (aock) [reply nav2_amcl/src/amcl_node.cpp:504]: Removed unnecessary mutex lock.

File: nav2_amcl/src/amcl_node.cpp (hunk 5 above)
Comment (aock) [reply nav2_amcl/src/amcl_node.cpp:1556]: Ah yes, I missed that. I added calling the reset function on cleanup like the others

**Comment by aock** [comment]
> also, heed the DCO requirement in the CI badges

Now the "Signed-off-by" is on a seperate line

File: nav2_amcl/src/amcl_node.cpp [hunk 6]
```diff
@@ -1543,6 +1553,10 @@ AmclNode::initServices()
...
+    "set_initial_pose",
+    std::bind(&AmclNode::initialGuessCallback, this, _1, _2, _3));
```
Comment (SteveMacenski) [inline nav2_amcl/src/amcl_node.cpp:1558]: This does not compile, that function does not exist anymore.

Please test your changes

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: nav2_amcl/src/amcl_node.cpp (hunk 6 above)
Comment (aock) [reply nav2_amcl/src/amcl_node.cpp]: Now it compiles again. At least on my computer

**Review by SteveMacenski — APPROVED** [review]

**Pull request merged**